### PR TITLE
[FLINK]add max idle time  config to prevent too frequent for empty commit

### PR DESCRIPTION
**Pull request by wg1026688210**
a draft for adding a  commit idle time config

### CODE REVIEW

**Comment by rdblue** [comment]
@openinx, could you review this?

**Comment by dixingxing0** [comment]
There is another new report on this issue: https://github.com/apache/iceberg/issues/2404 , should we  fix this before 0.12 release?

File: flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java [hunk 1]
```diff
@@ -154,10 +159,13 @@ public void snapshotState(StateSnapshotContext context) throws Exception {
     super.snapshotState(context);
     long checkpointId = context.getCheckpointId();
     LOG.info("Start to flush snapshot state to state backend, table: {}, checkpointId: {}", table, checkpointId);
+    long current = System.currentTimeMillis();
 
-    // Update the checkpoint state.
-    dataFilesPerCheckpoint.put(checkpointId, writeToManifest(checkpointId));
-
+    if (checkNeedCommit(current)) {
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:164]: If the `maxCommitIdleTimeMs` set to be 6min,   and the checkpoints produces such `WriteResult`: 

```
ckpt1:   {data-file1, data-file2},
ckpt2:   {},
ckpt3:   {data-file3, data-fille4},
ckpt4:   {},
```

If the checkpoint interval is set to be 3min,  then for every 6 min, we will still commit a dummy txn which has no data files in iceberg table ?   I mean currently we will periodily check whether the `WriteResult` is empty or not, if true then we will commit a dummy iceberg txn to refresh the latest max-committed-txn-id.  In fact when there're some txn that has committed few files into iceberg in a `maxCommitIdleTimeMs` interval, then we don't have to commit those empty txn to iceberg, right ?

Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:164]: So I think we could check how many continuous empty checkpoint we've encountered,  if it has exceeded the given threshold we could just commit a dummy txn to iceberg:

```diff
diff --git a/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java b/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java
index c1d3440d..00845ccd 100644
--- a/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java
+++ b/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java
@@ -95,6 +95,7 @@ class IcebergFilesCommitter extends AbstractStreamOperator<Void>
   private transient Table table;
   private transient ManifestOutputFileFactory manifestOutputFileFactory;
   private transient long maxCommittedCheckpointId;
+  private transient int continuousEmptyCheckpoints;
 
   // There're two cases that we restore from flink checkpoints: the first case is restoring from snapshot created by the
   // same flink job; another case is restoring from snapshot created by another different job. For the second case, we
@@ -125,6 +126,7 @@ class IcebergFilesCommitter extends AbstractStreamOperator<Void>
     int attemptId = getRuntimeContext().getAttemptNumber();
     this.manifestOutputFileFactory = FlinkManifestUtil.createOutputFileFactory(table, flinkJobId, subTaskId, attemptId);
     this.maxCommittedCheckpointId = INITIAL_CHECKPOINT_ID;
+    this.continuousEmptyCheckpoints = 0;
 
     this.checkpointsState = context.getOperatorStateStore().getListState(STATE_DESCRIPTOR);
     this.jobIdState = context.getOperatorStateStore().getListState(JOB_ID_DESCRIPTOR);
@@ -204,10 +206,18 @@ class IcebergFilesCommitter extends AbstractStreamOperator<Void>
       manifests.addAll(deltaManifests.manifests());
     }
 
-    if (replacePartitions) {
-      replacePartitions(pendingResults, newFlinkJobId, checkpointId);
-    } else {
-      commitDeltaTxn(pendingResults, newFlinkJobId, checkpointId);
+    int totalFiles = pendingResults.values().stream()
+        .mapToInt(r -> r.dataFiles().length + r.deleteFiles().length).sum();
+    continuousEmptyCheckpoints = totalFiles == 0 ? continuousEmptyCheckpoints + 1 : 0;
+
+    if (totalFiles != 0 || continuousEmptyCheckpoints % 10 == 0) {
+      if (replacePartitions) {
+        replacePartitions(pendingResults, newFlinkJobId, checkpointId);
+      } else {
+        commitDeltaTxn(pendingResults, newFlinkJobId, checkpointId);
+      }
+
+      continuousEmptyCheckpoints = 0;
     }
 
     pendingMap.clear();
```

File: flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java [hunk 2]
```diff
@@ -120,7 +123,10 @@ public void initializeState(StateInitializationContext context) throws Exception
     // Open the table loader and load the table.
     this.tableLoader.open();
     this.table = tableLoader.loadTable();
-
+    final Map<String, String> properties = table.properties();
+    maxContinuousEmptyCommits = PropertyUtil.propertyAsInt(properties, MAX_CONTINUOUS_EMPTY_COMMITS, 10);
+    Preconditions.checkArgument(maxContinuousEmptyCommits > 0, MAX_CONTINUOUS_EMPTY_COMMITS + " must not be " +
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:128]: As we will `MOD` the `maxContinuousEmptyCommits` in [here](https://github.com/apache/iceberg/pull/2042/files#diff-5320ef0a19dedc7a2693ccb49027baef1ad1a480df9d05688bbfe66410ce5bfcR214), so the `maxContinuousEmptyCommits` must be positive I think.  We may need to correct the error message in `Preconditions.checkArgument`

File: flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java [hunk 3]
```diff
@@ -120,7 +123,10 @@ public void initializeState(StateInitializationContext context) throws Exception
     // Open the table loader and load the table.
     this.tableLoader.open();
     this.table = tableLoader.loadTable();
-
+    final Map<String, String> properties = table.properties();
+    maxContinuousEmptyCommits = PropertyUtil.propertyAsInt(properties, MAX_CONTINUOUS_EMPTY_COMMITS, 10);
```
Comment (openinx) [inline flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:127]: We'd like to make this `maxContinuousEmptyCommits`'s assignment & validations into a separate block, that makes the code more clear.

```diff
diff --git a/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java b/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java
index 2fe6ab9e..56d25197 100644
--- a/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java
+++ b/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java
@@ -123,10 +123,11 @@ class IcebergFilesCommitter extends AbstractStreamOperator<Void>
     // Open the table loader and load the table.
     this.tableLoader.open();
     this.table = tableLoader.loadTable();
-    final Map<String, String> properties = table.properties();
-    maxContinuousEmptyCommits = PropertyUtil.propertyAsInt(properties, MAX_CONTINUOUS_EMPTY_COMMITS, 10);
-    Preconditions.checkArgument(maxContinuousEmptyCommits > 0, MAX_CONTINUOUS_EMPTY_COMMITS + " must not be " +
-        "negative");
+
+    this.maxContinuousEmptyCommits = PropertyUtil.propertyAsInt(table.properties(), MAX_CONTINUOUS_EMPTY_COMMITS, 10);
+    Preconditions.checkArgument(maxContinuousEmptyCommits > 0,
+        MAX_CONTINUOUS_EMPTY_COMMITS + " must not be negative");
+
     int subTaskId = getRuntimeContext().getIndexOfThisSubtask();
     int attemptId = getRuntimeContext().getAttemptNumber();
     this.manifestOutputFileFactory = FlinkManifestUtil.createOutputFileFactory(table, flinkJobId, subTaskId, attemptId);
```

File: flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java (hunk 2 above)
Comment (wg1026688210) [reply flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java:128]: ok,it's right

File: flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java [hunk 4]
```diff
@@ -140,6 +141,41 @@ public void testCommitTxnWithoutDataFiles() throws Exception {
...
+  }
+
+  private void doEmptyCommit(
+      OneInputStreamOperatorTestHarness<WriteResult, Void> harness,
+      long checkpointId,
+      long timestamp) throws Exception {
+    harness.setup();
+    harness.open();
+    harness.processElement(WriteResult.builder().build(), ++timestamp);
+    harness.snapshot(++checkpointId, ++timestamp);
```
Comment (openinx) [inline flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:167]: We iceberg don't allow to assign new value to the parameter `checkpointId` & `timestamp` in checkstyle, so I think that's why the travis checks has been broken ?

File: flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java [hunk 5]
```diff
@@ -140,6 +141,41 @@ public void testCommitTxnWithoutDataFiles() throws Exception {
...
+    harness.notifyOfCompletedCheckpoint(checkpointId);
+  }
+
+  private int snapshotCount() {
```
Comment (openinx) [inline flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:171]: We've already provided the `assertSnapshotSize`  so I think we don't have to introduce an extra `snapshotCount()` method now.

File: flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java [hunk 6]
```diff
@@ -140,6 +141,41 @@ public void testCommitTxnWithoutDataFiles() throws Exception {
     }
   }
 
+  @Test
+  public void testEmptyCommits() throws Exception {
```
Comment (openinx) [inline flink/src/test/java/org/apache/iceberg/flink/sink/TestIcebergFilesCommitter.java:145]: For this tests, it's good to simplify it as: 

```java
@Test
  public void testMaxContinuousEmptyCommits() throws Exception {
    table.updateProperties()
        .set(IcebergFilesCommitter.MAX_CONTINUOUS_EMPTY_COMMITS, "3")
        .commit();

    JobID jobId = new JobID();
    long checkpointId = 0;
    long timestamp = 0;
    try (OneInputStreamOperatorTestHarness<WriteResult, Void> harness = createStreamSink(jobId)) {
      harness.setup();
      harness.open();

      assertSnapshotSize(0);

      for (int i = 1; i <= 9; i++) {
        harness.snapshot(++checkpointId, ++timestamp);
        harness.notifyOfCompletedCheckpoint(checkpointId);

        assertSnapshotSize(i / 3);
      }
    }
  }
```

**Comment by openinx** [comment]
@wg1026688210 , could you pls check the checkstyle issue ?  https://github.com/apache/iceberg/runs/2284070477

File: flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java (hunk 3 above)
Comment (openinx) [reply flink/src/main/java/org/apache/iceberg/flink/sink/IcebergFilesCommitter.java]: @wg1026688210 could pls take a look at this minor comment ?

**Comment by openinx** [comment]
Got this merged,  thanks @wg1026688210 for contributing.

**Comment by wg1026688210** [comment]
> Got this merged, thanks @wg1026688210 for contributing.

thanks for your support

**Pull request merged**